src/wallet/wallet.cpp

```diff
@@ -260,6 +260,66 @@ std::shared_ptr<CWallet> LoadWalletInternal(WalletContext& context, const std::s
         return nullptr;
     }
 }
+
+class FastWalletRescanFilter
+{
+public:
+    FastWalletRescanFilter(const CWallet& wallet) : m_wallet(wallet)
+    {
+        // fast rescanning via block filters is only supported by descriptor wallets right now
+        assert(!m_wallet.IsLegacy());
+
+        // create initial filter with scripts from both active and non-active ScriptPubKeyMans
+        auto active_spkms = m_wallet.GetActiveScriptPubKeyMans();
+        for (auto spkm : m_wallet.GetAllScriptPubKeyMans()) {
+            auto desc_spkm = dynamic_cast<DescriptorScriptPubKeyMan*>(spkm);
+            for (const auto& script_pub_key : desc_spkm->GetScriptPubKeys()) {
+                m_filter_set.emplace(script_pub_key.begin(), script_pub_key.end());
+            }
+            // save active ScriptPubKeyMans's end ranges for possible future filter updates
+            if (active_spkms.find(spkm) != active_spkms.end()) {
+                m_last_range_ends.emplace(desc_spkm, GetDescriptorRangeEnd(*desc_spkm));
+            }
+        }
+    }
+
+    void UpdateIfNeeded()
+    {
+        // repopulate filter with new scripts if top-up has happened since last iteration
+        for (auto [desc_spkm, last_range_end] : m_last_range_ends) {
```

Comment:
Not needed. Every import command checks if a rescan is being performed prior execution (it's the `WalletRescanReserver` purpose).

Reply:
Thanks for clarifying. At first sight, that seems quite a fragile assumption since it's implicit and requires every user of FastWalletRescanFilter to do that check voluntarily. Some kind of assertion would be nice but would go at the cost of performance to call and cast all spkms after every block, so maybe not worth it.